src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PlaceholderReplacer.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp;
+
+internal sealed partial class LocalRewriter
+{
+    private sealed class PlaceholderReplacer : BoundTreeRewriterWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+    {
+        private readonly Dictionary<BoundValuePlaceholderBase, BoundExpression> _placeholders;
+
+        PlaceholderReplacer(Dictionary<BoundValuePlaceholderBase, BoundExpression> placeholders)
+        {
+            _placeholders = placeholders;
+        }
+
+        public static BoundExpression Replace(Dictionary<BoundValuePlaceholderBase, BoundExpression> placeholders, BoundExpression expr)
```

Comment:
>BoundValuePlaceholderBase [](http://example.com/codeflow?start=57&length=25)

Perhaps it would be worth introducing a common base specifically for list/slice pattern placeholders so that we could make this dictionary stronger typed.